README.md
# Sensor Unit Normalizer

Convert sensor readings between units of the same dimension, with dimension
checks, physical ranges and explicit handling of values that are not readings.

- **Repository:** [edilec/sensor-unit-normalizer](https://github.com/edilec/sensor-unit-normalizer)
- **Area:** IoT & Edge
- **License:** MIT

## Why it exists

Two failures show up again and again in telemetry pipelines.

The first is treating an **affine** conversion as a linear one. Scaling 20 °C by
9/5 gives 36 instead of 68, and a registry that only knows about multipliers
will do exactly that without complaining.

The second is **`NaN`**. It propagates silently through arithmetic, so one bad
sample can poison an entire derived series while every downstream check still
reports success.

This tool handles both at the boundary.

## Requirements

- Node.js 22 or newer
- no runtime dependencies

## Quick start

```sh
node bin/sensor-unit-normalizer.mjs convert 20 degC degF --precision 2   # 68
node bin/sensor-unit-normalizer.mjs normalize examples/readings.json --precision 3
node bin/sensor-unit-normalizer.mjs units
```

A batch records why each rejected reading failed, and still normalizes the rest:

```text
ERROR   /1 Cannot convert m (length) to kg (mass).
ERROR   /2 -400 degC is outside the physically possible range for temperature.
ERROR   /3 Unknown source unit "furlong".
ERROR   /4 No target unit declared for reading no-target.

5 reading(s): 1 normalized, 4 rejected, status fail.
```

A single bad sample in a telemetry export should not cost you the whole file.

## Library usage

```js
import { convert, normalizeReadings } from 'sensor-unit-normalizer'

convert(20, 'degC', 'degF')                    // 67.99999999999999
convert(20, 'degC', 'degF', { precision: 2 })  // 68
convert(1, 'm', 'kg')                          // throws UnitError dimension-mismatch
```

## Floating point

Conversions run in IEEE-754 doubles, and an affine conversion composes a
multiplication with an offset, so results carry roughly one unit in the last
place of error — `convert(0, 'degC', 'degF')` is `31.999999999999986`, not
exactly `32`.

That is expected, and it is why round-trip guarantees are stated as *within
tolerance*. Pass `precision` when you need an exact decimal for display or for a
stable serialized record.

## Units

Eleven dimensions: temperature, length, mass, time, pressure, speed, electric
potential, electric current, data size, angle and relative humidity. Decimal and
binary data prefixes are deliberately distinct — `GB` is 10⁹ bytes, `GiB` is 2³⁰.

The full registry, ranges and rule catalog are in
[`docs/units.md`](./docs/units.md).

## Exit codes

| Code | Meaning |
| ---: | --- |
| `0` | every reading converted |
| `1` | at least one reading could not be converted |
| `2` | invalid usage, or the input could not be read |

## Limits and non-goals

- It converts units; it does not calibrate sensors, correct drift, or judge
  whether a reading is plausible beyond the coarse physical range.
- Ranges are sanity bounds on real-world readings, not calibration limits. Use
  `--no-range-check` for calibration or fault-injection data.
- It does not parse compound or derived units: `m/s` is a registered unit, not
  something assembled from `m` and `s`.
- It does not carry measurement uncertainty. A converted value is as precise as
  the double arithmetic that produced it, no more.
- Relative humidity is treated as a plain percentage; it is not converted to or
  from absolute humidity, which needs temperature and pressure.

## Development

```sh
npm test          # behaviour tests
npm run check     # lint, tests, runnable example, packaging check
```

## License

MIT. See [LICENSE](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project are documented in this file.

## Unreleased

### Added

- a typed unit registry covering eleven dimensions, with affine conversion
  support so Celsius, Fahrenheit and Kelvin convert across their offsets rather
  than by scaling alone;
- dimension checking, so incompatible units are rejected instead of silently
  scaled;
- physical range checks per dimension, with an opt-out for calibration data;
- rejection of `NaN`, infinities and non-numbers at the boundary;
- `precision` rounding that avoids the usual `toFixed` surprises;
- `normalizeReadings` for batches, recording why each rejected reading failed
  while still normalizing the rest;
- a CLI with `convert`, `normalize` and `units`, and exit codes 0 / 1 / 2;
- runnable clean and broken reading examples;
- the registry and conversion rules in `docs/units.md`.

### Fixed

- `normalize` no longer republishes the readings file it could not parse. V8
  reports a parse failure two ways, and one of them quotes the input back --
  `Unexpected token 'A', "AKIAIOSFODNN7EXAMPLE" is not valid JSON` -- which
  reproduces the first ten characters of the file, or the whole file when it is
  shorter than that. The command line let that message reach stderr whole, so a
  readings file short enough to be nothing but a credential was printed in full.
  `parseFailureDetail` in `src/index.mjs` now keeps the offset, line and column
  and drops the quoted half. `test/parse-failure.test.mjs` plants the canary
  through the real binary and asserts it is absent from stdout, from stderr and
  from every prefix of it down to eight characters, because V8 quotes only ten.

No release has been published.


test/cli.test.mjs
import assert from 'node:assert/strict'
import { spawnSync } from 'node:child_process'
import { dirname, resolve } from 'node:path'
import test from 'node:test'
import { fileURLToPath } from 'node:url'

const projectDirectory = resolve(dirname(fileURLToPath(import.meta.url)), '..')
const cli = resolve(projectDirectory, 'bin/sensor-unit-normalizer.mjs')

function run(args) {
  return spawnSync(process.execPath, [cli, ...args], { cwd: projectDirectory, encoding: 'utf8' })
}

test('convert prints the converted value', () => {
  const result = run(['convert', '20', 'degC', 'degF', '--precision', '2'])

  assert.equal(result.status, 0)
  assert.equal(result.stdout.trim(), '68')
})

test('convert refuses an incompatible pair with a non-zero exit', () => {
  const result = run(['convert', '1', 'm', 'kg'])

  assert.notEqual(result.status, 0)
  assert.equal(result.stdout, '')
  assert.match(result.stderr, /Cannot convert m \(length\) to kg \(mass\)/)
})

test('convert refuses NaN input', () => {
  const result = run(['convert', 'not-a-number', 'degC', 'degF'])

  assert.notEqual(result.status, 0)
  assert.match(result.stderr, /NaN/)
})

test('normalize on the clean example exits zero', () => {
  const result = run(['normalize', 'examples/readings.json', '--precision', '3'])

  assert.equal(result.status, 0)
  assert.match(result.stdout, /status pass/)
})

test('normalize on the broken example exits one and names each reason', () => {
  const result = run(['normalize', 'examples/broken-readings.json'])

  assert.equal(result.status, 1)
  assert.match(result.stdout, /Cannot convert m \(length\) to kg \(mass\)/)
  assert.match(result.stdout, /physically possible range/)
  assert.match(result.stdout, /Unknown source unit "furlong"/)
})

test('--json is parseable with a clean stderr', () => {
  const result = run(['normalize', 'examples/readings.json', '--json'])

  assert.equal(result.stderr, '')
  assert.equal(JSON.parse(result.stdout).tool, 'sensor-unit-normalizer')
})

test('repeated runs produce identical JSON', () => {
  const args = ['normalize', 'examples/broken-readings.json', '--json']
  assert.equal(run(args).stdout, run(args).stdout)
})

test('units lists the registry', () => {
  const result = run(['units'])
  const names = result.stdout.trim().split('\n')

  assert.equal(result.status, 0)
  assert.ok(names.includes('degC'))
  assert.ok(names.includes('GiB'))
  assert.deepEqual(names, [...names].sort())
})

test('--no-range-check allows calibration values', () => {
  const blocked = run(['convert', '-300', 'degC', 'degC'])
  const allowed = run(['convert', '-300', 'degC', 'degC', '--no-range-check'])

  assert.notEqual(blocked.status, 0)
  assert.equal(allowed.status, 0)
})

test('a missing file and unknown command are usage errors', () => {
  assert.equal(run(['normalize', 'examples/nope.json']).status, 2)
  assert.match(run(['frobnicate']).stderr, /Unknown command "frobnicate"/)
})

test('--help exits zero', () => {
  const result = run(['--help'])

  assert.equal(result.status, 0)
  assert.match(result.stdout, /Usage:/)
})


test/convert.test.mjs
import assert from 'node:assert/strict'
import test from 'node:test'

import {
  DIMENSIONS,
  UNITS,
  UnitError,
  areCompatible,
  convert,
  dimensionOf,
  normalizeReadings,
  round,
  unitNames,
} from '../src/index.mjs'

const TOLERANCE = 1e-9

function close(actual, expected, message) {
  assert.ok(
    Math.abs(actual - expected) < TOLERANCE,
    `${message ?? ''} expected ${expected}, got ${actual} (delta ${Math.abs(actual - expected)})`,
  )
}

test('Celsius and Fahrenheit convert across their offset, not by scaling alone', () => {
  // The classic failure is a linear-only registry: 20 * 9/5 = 36 instead of 68.
  close(convert(0, 'degC', 'degF'), 32, '0C')
  close(convert(20, 'degC', 'degF'), 68, '20C')
  close(convert(100, 'degC', 'degF'), 212, '100C')
  close(convert(-40, 'degC', 'degF'), -40, '-40C is the crossover')
  close(convert(98.6, 'degF', 'degC'), 37, 'body temperature')
})

test('Celsius/Fahrenheit round-trips stay within tolerance', () => {
  for (let celsius = -273; celsius <= 1000; celsius += 7.5) {
    const roundTripped = convert(convert(celsius, 'degC', 'degF'), 'degF', 'degC')
    close(roundTripped, celsius, `round trip at ${celsius}C`)
  }
})

test('Kelvin is the temperature base and anchors both scales', () => {
  close(convert(0, 'K', 'degC'), -273.15)
  close(convert(273.15, 'K', 'degC'), 0)
  close(convert(0, 'K', 'degF'), -459.67)
})

test('every unit round-trips through its dimension base within tolerance', () => {
  for (const name of unitNames()) {
    const base = DIMENSIONS[UNITS[name].dimension].base
    const sample = name === '%RH' ? 50 : 1
    const roundTripped = convert(convert(sample, name, base), base, name)
    close(roundTripped, sample, `${name} round trip`)
  }
})

test('incompatible dimensions are rejected, never silently scaled', () => {
  assert.throws(() => convert(1, 'm', 'kg'), (error) => {
    assert.ok(error instanceof UnitError)
    assert.equal(error.code, 'dimension-mismatch')
    assert.equal(error.fromDimension, 'length')
    assert.equal(error.toDimension, 'mass')
    return true
  })
  assert.throws(() => convert(1, 'degC', 's'), /Cannot convert degC \(temperature\) to s \(time\)/)
})

test('NaN is refused rather than propagated', () => {
  assert.throws(() => convert(Number.NaN, 'degC', 'degF'), (error) => {
    assert.equal(error.code, 'value-not-a-number')
    assert.match(error.message, /NaN/)
    return true
  })
})

test('infinities and non-numbers are refused', () => {
  assert.throws(() => convert(Number.POSITIVE_INFINITY, 'm', 'km'), /value is Infinity|is Infinity/i)
  assert.throws(() => convert(Number.NEGATIVE_INFINITY, 'm', 'km'), /Infinity/)
  assert.throws(() => convert('20', 'degC', 'degF'), /must be a number/)
  assert.throws(() => convert(null, 'degC', 'degF'), /received null/)
  assert.throws(() => convert(undefined, 'degC', 'degF'), /must be a number/)
})

test('an unknown unit names itself and its role', () => {
  assert.throws(() => convert(1, 'furlong', 'm'), /Unknown source unit "furlong"/)
  assert.throws(() => convert(1, 'm', 'furlong'), /Unknown target unit "furlong"/)
})

test('physically impossible readings are rejected', () => {
  assert.throws(() => convert(-300, 'degC', 'K'), (error) => {
    assert.equal(error.code, 'value-out-of-range')
    assert.equal(error.dimension, 'temperature')
    return true
  })
  assert.throws(() => convert(-1, 'm', 'cm'), /out of range|physically possible/)
  assert.throws(() => convert(101, '%RH', '%RH'), /physically possible/)
})

test('range checking can be turned off for a calibration workflow', () => {
  close(convert(-300, 'degC', 'degC', { checkRange: false }), -300)
})

test('decimal and binary data prefixes stay distinct', () => {
  close(convert(1, 'GiB', 'MiB'), 1024)
  close(convert(1, 'GB', 'MB'), 1000)
  assert.notEqual(convert(1, 'GiB', 'B'), convert(1, 'GB', 'B'))
})

test('known reference conversions are correct', () => {
  close(convert(1, 'mi', 'm'), 1609.344)
  close(convert(1, 'lb', 'kg'), 0.45359237)
  close(convert(1, 'bar', 'psi'), 14.503773773020924, 'bar to psi')
  close(convert(180, 'deg', 'rad'), Math.PI)
  close(convert(1, 'h', 's'), 3600)
  close(convert(100, 'km/h', 'm/s'), 27.77777777777778)
})

test('precision rounds without the usual toFixed surprises', () => {
  assert.equal(round(1.005, 2), 1.01)
  assert.equal(round(2.675, 2), 2.68)
  assert.equal(convert(0, 'degC', 'degF', { precision: 6 }), 32)
  assert.equal(convert(1, 'bar', 'psi', { precision: 2 }), 14.5)
})

test('an invalid precision is rejected', () => {
  assert.throws(() => round(1, -1), /integer from 0 through 15/)
  assert.throws(() => round(1, 1.5), /integer from 0 through 15/)
  assert.throws(() => round(1, 16), /integer from 0 through 15/)
})

test('dimension helpers report compatibility', () => {
  assert.equal(dimensionOf('degF'), 'temperature')
  assert.equal(areCompatible('m', 'ft'), true)
  assert.equal(areCompatible('m', 'kg'), false)
  assert.equal(areCompatible('m', 'furlong'), false)
})

test('a batch normalizes the good readings and records why the rest failed', () => {
  const report = normalizeReadings([
    { id: 'a', value: 20, unit: 'degC', targetUnit: 'degF' },
    { id: 'b', value: Number.NaN, unit: 'degC', targetUnit: 'degF' },
    { id: 'c', value: 1, unit: 'm', targetUnit: 'kg' },
    { id: 'd', value: 1, unit: 'km', targetUnit: 'm' },
  ], { precision: 4 })

  assert.equal(report.status, 'fail')
  assert.equal(report.summary.normalized, 2)
  assert.equal(report.summary.errors, 2)
  assert.equal(report.readings[0].value, 68)
  assert.equal(report.readings[0].originalUnit, 'degC')
  assert.deepEqual(report.findings.map((item) => item.ruleId), ['value-not-a-number', 'dimension-mismatch'])
})

test('a clean batch passes with the report envelope', () => {
  const report = normalizeReadings([{ value: 1, unit: 'km', targetUnit: 'm' }])

  assert.equal(report.schemaVersion, '1')
  assert.equal(report.tool, 'sensor-unit-normalizer')
  assert.equal(report.status, 'pass')
  assert.deepEqual(report.findings, [])
})

test('a per-dimension target unit applies to every reading of that dimension', () => {
  const report = normalizeReadings(
    [{ value: 20, unit: 'degC' }, { value: 1, unit: 'km' }],
    { targetUnit: { temperature: 'K', length: 'm' }, precision: 2 },
  )

  assert.equal(report.status, 'pass')
  assert.equal(report.readings[0].value, 293.15)
  assert.equal(report.readings[1].value, 1000)
})

test('a reading with no target unit is reported, not guessed', () => {
  const report = normalizeReadings([{ value: 1, unit: 'km' }])

  assert.deepEqual(report.findings.map((item) => item.ruleId), ['target-unit-missing'])
})

test('a malformed reading is reported without crashing the batch', () => {
  const report = normalizeReadings(['nope', null, { value: 1, unit: 'km', targetUnit: 'm' }])

  assert.equal(report.summary.normalized, 1)
  assert.equal(report.findings.filter((item) => item.ruleId === 'reading-malformed').length, 2)
})

test('batch results are deterministic', () => {
  const input = [
    { value: Number.NaN, unit: 'degC', targetUnit: 'degF' },
    { value: 1, unit: 'm', targetUnit: 'kg' },
  ]

  assert.deepEqual(normalizeReadings(input), normalizeReadings(input))
})

test('a non-array batch is rejected', () => {
  assert.throws(() => normalizeReadings(null), /must be an array/)
})


test/parse-failure.test.mjs
import assert from 'node:assert/strict'
import { spawnSync } from 'node:child_process'
import { mkdtemp, rm, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { dirname, join, resolve } from 'node:path'
import test from 'node:test'
import { fileURLToPath } from 'node:url'

import { parseFailureDetail } from '../src/index.mjs'

/**
 * A readings file that cannot be parsed must not be republished by its own
 * diagnostic.
 *
 * V8 reports a parse failure two ways, and one of them quotes the input back:
 * `Unexpected token 'A', "AKIAIOSFODNN7EXAMPLE" is not valid JSON`. The quoted
 * run is the first ten characters of the document, or the whole document when
 * it is shorter, so a file short enough to be nothing but a credential is
 * reproduced in full by its own error message. `normalize` interpolated that
 * message whole onto stderr -- on the error path, which is the path a
 * malformed or misnamed file is guaranteed to take.
 *
 * The canary is `AKIAIOSFODNN7EXAMPLE`, the access key id AWS publishes in its
 * own documentation. It is not a credential; it is the shape of one.
 */

const projectDirectory = resolve(dirname(fileURLToPath(import.meta.url)), '..')
const cli = resolve(projectDirectory, 'bin/sensor-unit-normalizer.mjs')

const CANARY = 'AKIAIOSFODNN7EXAMPLE'

function run(args) {
  return spawnSync(process.execPath, [cli, ...args], { cwd: projectDirectory, encoding: 'utf8' })
}

/** Write a readings file of the caller's choosing and run the real binary over it. */
async function withReadings(content, body) {
  const base = await mkdtemp(join(tmpdir(), 'sensor-unit-normalizer-canary-'))
  try {
    const path = join(base, 'readings.json')
    await writeFile(path, content, 'utf8')
    return await body(run(['normalize', path]))
  } finally {
    await rm(base, { recursive: true, force: true })
  }
}

/**
 * Assert the canary is absent from both streams, and so is every prefix of it
 * down to eight characters. The prefixes matter because V8 quotes only the
 * first ten characters once the document is long enough: a test looking for
 * the whole canary would pass against a message still leaking `AKIAIOSFOD`.
 */
function assertNoCanary(result, label) {
  for (let length = CANARY.length; length >= 8; length -= 1) {
    const prefix = CANARY.slice(0, length)
    assert.equal(result.stdout.includes(prefix), false, `${label}: stdout carries ${prefix}`)
    assert.equal(result.stderr.includes(prefix), false, `${label}: stderr carries ${prefix}`)
  }
}

test('a readings file that is nothing but a credential is not echoed back', async () => {
  await withReadings(CANARY, (result) => {
    assert.equal(result.status, 2)
    assert.equal(result.stdout, '')
    assert.match(result.stderr, /not valid JSON/)
    assertNoCanary(result, 'whole document')
  })
})

test('a credential inside a longer readings file is not echoed either', async () => {
  await withReadings(`{"targetUnit": "degC", "readings": ${CANARY}}`, (result) => {
    assert.equal(result.status, 2)
    assertNoCanary(result, 'embedded in a document')
  })
})

test('the position, line and column survive -- a diagnostic that says nothing is a different defect', async () => {
  await withReadings('{"targetUnit": "degC" "readings": []}', (result) => {
    assert.equal(result.status, 2)
    assert.match(result.stderr, /at position \d+ \(line \d+ column \d+\)/)
  })
})

test('a readings file that cannot be read still says which syscall failed', () => {
  const result = run(['normalize', 'examples/nope.json'])

  assert.equal(result.status, 2)
  assert.match(result.stderr, /ENOENT/)
})

test('parseFailureDetail keeps the offset and drops the quoted document', () => {
  const detailFor = (text) => {
    try {
      JSON.parse(text)
    } catch (error) {
      return parseFailureDetail(error)
    }
    throw new Error('the fixture parsed, so it pins nothing')
  }

  assert.equal(detailFor(CANARY), "unexpected token 'A' at the start of the document")
  assert.equal(detailFor('ssn 123-45-6789'), "unexpected token 's' at the start of the document")
  assert.equal(detailFor('password=hunter2-correct-horse'), "unexpected token 'p' at the start of the document")
  assert.equal(detailFor(`{"a": 1, "b": ${CANARY}}`), "unexpected token 'A' inside the document")
  assert.equal(detailFor(''), 'Unexpected end of JSON input')
  assert.equal(detailFor('{"a": 1 "b": 2}'), "Expected ',' or '}' after property value in JSON at position 8 (line 1 column 9)")
  assert.equal(detailFor('{"a": 1} trailing'), 'Unexpected non-whitespace character after JSON at position 9 (line 1 column 10)')
})

test('parseFailureDetail refuses a wording it was not taught rather than guessing', () => {
  assert.equal(
    parseFailureDetail(new Error('Unexpected token \'A\', "AKIAIOSFODNN7EXAMPLE" is not valid JSON at position 0')),
    'the document could not be parsed as JSON',
    'a double quote surviving to the end means the snippet survived with it',
  )
  assert.equal(parseFailureDetail(undefined), 'the document could not be parsed as JSON')
})

/**
 * The order the two V8 spellings are recognised in is what keeps the rest true.
 *
 * `parseFailureDetail` looks for the quoting shape BEFORE it looks for the
 * offset, and the difference is not cosmetic: a readings file whose own text
 * reads `at position 1` is quoted back by V8 as `Unexpected token 'a', "at
 * position 1" is not valid JSON`, so an offset-first search finds that phrase
 * inside the quoted span and the slice hands the file straight back out.
 * Nineteen tools in this catalog shipped that ordering. These tests fail if
 * this one adopts it.
 */

const UNPARSEABLE = 'the document could not be parsed as JSON'

/** The detail for a document V8 actually refused -- never a hand-written message. */
function detailOf(document) {
  try {
    JSON.parse(document)
  } catch (error) {
    return parseFailureDetail(error)
  }
  throw new Error(`${JSON.stringify(document)} parsed, so it pins nothing`)
}

/** No run of four characters or more from the document survives into the detail. */
function assertNoRunOf(document, detail, label) {
  for (let length = 4; length <= document.length; length += 1) {
    const prefix = document.slice(0, length)
    assert.equal(detail.includes(prefix), false, `${label}: the detail carries ${JSON.stringify(prefix)}`)
  }
}

test('a readings file whose own text reads "at position 1" is not sliced back out', () => {
  const document = 'at position 1'
  const detail = detailOf(document)

  assert.equal(detail.includes('"'), false, 'a double quote in the detail means a quoted span survived')
  assert.equal(detail.includes(document), false, 'the file came back inside its own diagnostic')
  assert.equal(detail, "unexpected token 'a' at the start of the document")
})

test('a long readings file is not quoted back by any run of its sensitive opening', () => {
  const document = `${CANARY} and a great deal of trailing content nobody should read back`
  const detail = detailOf(document)

  assertNoRunOf(document, detail, 'long readings file')
  assert.equal(detail, "unexpected token 'A' at the start of the document")
})

test('a quoted span holding a newline is still recognised as a quoted span', () => {
  // Without the `s` flag the quoting shape does not match a span with a line
  // break in it, the offset branch is reached, and the generic sentence is all
  // that is left -- or worse, the span is sliced out.
  const detail = detailOf('}x\n')

  assert.equal(detail, "unexpected token '}' at the start of the document")
  assert.notEqual(detail, UNPARSEABLE, 'the newline case fell through to the generic sentence')
})

test('the safe positional spelling keeps its position, line and column', () => {
  // A helper that answered the generic sentence for everything would pass every
  // leak test above while destroying every diagnostic. This is the other half.
  const detail = detailOf('{"targetUnit": "degC" "readings": []}')

  assert.match(detail, /at position \d+ \(line \d+ column \d+\)$/)
  assert.equal(detail, "Expected ',' or '}' after property value in JSON at position 22 (line 1 column 23)")
})


src/units.mjs
/**
 * The unit registry.
 *
 * Every unit converts to its dimension's base unit by `value * factor + offset`.
 * Most units are purely multiplicative; temperature is not, which is the whole
 * reason `offset` exists. Getting that wrong is the classic sensor bug: scaling
 * 20 °C by 9/5 gives 36 instead of 68, and a linear-only registry will do
 * exactly that without complaining.
 *
 * `range` is the physically possible span expressed in the base unit. It is a
 * sanity bound on real-world readings, not a calibration limit.
 */

export const DIMENSIONS = Object.freeze({
  temperature: { base: 'K', range: { min: 0, max: 1e9 } },
  length: { base: 'm', range: { min: 0, max: 1e15 } },
  mass: { base: 'kg', range: { min: 0, max: 1e12 } },
  time: { base: 's', range: { min: 0, max: 1e12 } },
  pressure: { base: 'Pa', range: { min: 0, max: 1e12 } },
  speed: { base: 'm/s', range: { min: 0, max: 299_792_458 } },
  'electric-potential': { base: 'V', range: { min: -1e9, max: 1e9 } },
  'electric-current': { base: 'A', range: { min: -1e9, max: 1e9 } },
  'data-size': { base: 'B', range: { min: 0, max: 1e21 } },
  angle: { base: 'rad', range: { min: -1e6, max: 1e6 } },
  'relative-humidity': { base: '%RH', range: { min: 0, max: 100 } },
})

/** unit -> { dimension, factor, offset } into the dimension's base unit. */
export const UNITS = Object.freeze({
  // temperature: the affine cases
  K: { dimension: 'temperature', factor: 1, offset: 0 },
  degC: { dimension: 'temperature', factor: 1, offset: 273.15 },
  degF: { dimension: 'temperature', factor: 5 / 9, offset: 273.15 - (32 * 5) / 9 },

  // length
  m: { dimension: 'length', factor: 1, offset: 0 },
  km: { dimension: 'length', factor: 1000, offset: 0 },
  cm: { dimension: 'length', factor: 0.01, offset: 0 },
  mm: { dimension: 'length', factor: 0.001, offset: 0 },
  mi: { dimension: 'length', factor: 1609.344, offset: 0 },
  ft: { dimension: 'length', factor: 0.3048, offset: 0 },
  in: { dimension: 'length', factor: 0.0254, offset: 0 },

  // mass
  kg: { dimension: 'mass', factor: 1, offset: 0 },
  g: { dimension: 'mass', factor: 0.001, offset: 0 },
  mg: { dimension: 'mass', factor: 1e-6, offset: 0 },
  lb: { dimension: 'mass', factor: 0.45359237, offset: 0 },

  // time
  s: { dimension: 'time', factor: 1, offset: 0 },
  ms: { dimension: 'time', factor: 0.001, offset: 0 },
  min: { dimension: 'time', factor: 60, offset: 0 },
  h: { dimension: 'time', factor: 3600, offset: 0 },
  d: { dimension: 'time', factor: 86_400, offset: 0 },

  // pressure
  Pa: { dimension: 'pressure', factor: 1, offset: 0 },
  hPa: { dimension: 'pressure', factor: 100, offset: 0 },
  kPa: { dimension: 'pressure', factor: 1000, offset: 0 },
  bar: { dimension: 'pressure', factor: 100_000, offset: 0 },
  psi: { dimension: 'pressure', factor: 6894.757293168361, offset: 0 },

  // speed
  'm/s': { dimension: 'speed', factor: 1, offset: 0 },
  'km/h': { dimension: 'speed', factor: 1000 / 3600, offset: 0 },
  mph: { dimension: 'speed', factor: 1609.344 / 3600, offset: 0 },

  // electrical
  V: { dimension: 'electric-potential', factor: 1, offset: 0 },
  mV: { dimension: 'electric-potential', factor: 0.001, offset: 0 },
  A: { dimension: 'electric-current', factor: 1, offset: 0 },
  mA: { dimension: 'electric-current', factor: 0.001, offset: 0 },

  // data size: decimal and binary prefixes are distinct on purpose
  B: { dimension: 'data-size', factor: 1, offset: 0 },
  kB: { dimension: 'data-size', factor: 1000, offset: 0 },
  MB: { dimension: 'data-size', factor: 1e6, offset: 0 },
  GB: { dimension: 'data-size', factor: 1e9, offset: 0 },
  KiB: { dimension: 'data-size', factor: 1024, offset: 0 },
  MiB: { dimension: 'data-size', factor: 1024 ** 2, offset: 0 },
  GiB: { dimension: 'data-size', factor: 1024 ** 3, offset: 0 },

  // angle
  rad: { dimension: 'angle', factor: 1, offset: 0 },
  deg: { dimension: 'angle', factor: Math.PI / 180, offset: 0 },

  // relative humidity
  '%RH': { dimension: 'relative-humidity', factor: 1, offset: 0 },
})

export function unitNames() {
  return Object.keys(UNITS).sort()
}


src/index.mjs
import { DIMENSIONS, UNITS, unitNames } from './units.mjs'

export { DIMENSIONS, UNITS, unitNames } from './units.mjs'

export const TOOL_ID = 'sensor-unit-normalizer'
export const REPORT_SCHEMA_VERSION = '1'

/** Thrown for a conversion that cannot be performed at all. */
export class UnitError extends Error {
  constructor(code, message, details = {}) {
    super(message)
    this.name = 'UnitError'
    this.code = code
    Object.assign(this, details)
  }
}

function requireUnit(name, role) {
  const unit = Object.hasOwn(UNITS, name) ? UNITS[name] : undefined
  if (!unit) {
    throw new UnitError('unknown-unit', `Unknown ${role} unit "${name}".`, { unit: name })
  }
  return unit
}

/**
 * Reject anything that is not a real reading.
 *
 * NaN is the important one: it propagates silently through arithmetic, so a
 * single bad sample can poison an entire derived series while every downstream
 * check still "succeeds". It is refused at the boundary rather than converted.
 */
function requireFiniteNumber(value) {
  if (typeof value !== 'number') {
    throw new UnitError('value-not-a-number', `Value must be a number; received ${value === null ? 'null' : typeof value}.`, {
      valueType: value === null ? 'null' : typeof value,
    })
  }
  if (Number.isNaN(value)) {
    throw new UnitError('value-not-a-number', 'Value is NaN, which is not a reading.', { valueType: 'NaN' })
  }
  if (!Number.isFinite(value)) {
    throw new UnitError('value-not-finite', `Value is ${value > 0 ? 'Infinity' : '-Infinity'}, which is not a reading.`, {
      valueType: 'Infinity',
    })
  }
  return value
}

function toBase(value, unit) {
  return value * unit.factor + unit.offset
}

function fromBase(value, unit) {
  return (value - unit.offset) / unit.factor
}

/**
 * Convert a reading between units of the same dimension.
 *
 * @param {number} value
 * @param {string} from
 * @param {string} to
 * @param {{ precision?: number, checkRange?: boolean }} [options]
 */
export function convert(value, from, to, options = {}) {
  const source = requireUnit(from, 'source')
  const target = requireUnit(to, 'target')

  if (source.dimension !== target.dimension) {
    throw new UnitError(
      'dimension-mismatch',
      `Cannot convert ${from} (${source.dimension}) to ${to} (${target.dimension}).`,
      { from, to, fromDimension: source.dimension, toDimension: target.dimension },
    )
  }

  requireFiniteNumber(value)

  const base = toBase(value, source)

  if (options.checkRange !== false) {
    const { range } = DIMENSIONS[source.dimension]
    if (base < range.min || base > range.max) {
      throw new UnitError(
        'value-out-of-range',
        `${value} ${from} is outside the physically possible range for ${source.dimension}.`,
        {
          value,
          unit: from,
          dimension: source.dimension,
          baseValue: base,
          baseUnit: DIMENSIONS[source.dimension].base,
          range,
        },
      )
    }
  }

  const result = fromBase(base, target)

  if (!Number.isFinite(result)) {
    throw new UnitError('value-not-finite', 'Conversion produced a non-finite result.', { from, to, value })
  }

  return options.precision === undefined ? result : round(result, options.precision)
}

/**
 * Round to a number of decimal places without the usual floating-point
 * surprises of `toFixed` on values like 1.005.
 */
export function round(value, precision) {
  if (!Number.isInteger(precision) || precision < 0 || precision > 15) {
    throw new UnitError('invalid-precision', 'Precision must be an integer from 0 through 15.', { precision })
  }
  const scaled = Number(`${value}e${precision}`)
  if (!Number.isFinite(scaled)) return value
  return Number(`${Math.round(scaled)}e-${precision}`)
}

export function dimensionOf(unit) {
  return requireUnit(unit, 'source').dimension
}

export function areCompatible(from, to) {
  return Object.hasOwn(UNITS, from)
    && Object.hasOwn(UNITS, to)
    && UNITS[from].dimension === UNITS[to].dimension
}

function byCodeUnit(left, right) {
  if (left === right) return 0
  return left < right ? -1 : 1
}

/**
 * Normalize a batch of readings.
 *
 * One unconvertible reading does not abandon the batch: it is recorded with its
 * reason and the rest are still normalized, because a single bad sample in a
 * telemetry export should not cost you the whole file.
 */
export function normalizeReadings(readings, options = {}) {
  if (!Array.isArray(readings)) throw new TypeError('Readings must be an array')

  const normalized = []
  const findings = []

  readings.forEach((reading, index) => {
    const pointer = `/${index}`
    if (!reading || typeof reading !== 'object' || Array.isArray(reading)) {
      findings.push({
        ruleId: 'reading-malformed',
        severity: 'error',
        message: 'Reading must be an object.',
        location: { pointer },
      })
      return
    }

    const target = reading.targetUnit ?? options.targetUnit?.[
      Object.hasOwn(UNITS, reading.unit) ? UNITS[reading.unit].dimension : ''
    ]

    if (target === undefined) {
      findings.push({
        ruleId: 'target-unit-missing',
        severity: 'error',
        message: `No target unit declared for reading ${reading.id ?? index}.`,
        location: { pointer },
      })
      return
    }

    try {
      const value = convert(reading.value, reading.unit, target, {
        precision: options.precision,
        checkRange: options.checkRange,
      })
      normalized.push({
        ...(reading.id !== undefined ? { id: reading.id } : {}),
        value,
        unit: target,
        originalValue: reading.value,
        originalUnit: reading.unit,
      })
    } catch (error) {
      if (!(error instanceof UnitError)) throw error
      findings.push({
        ruleId: error.code,
        severity: 'error',
        message: error.message,
        location: { pointer },
        ...(reading.id !== undefined ? { id: reading.id } : {}),
      })
    }
  })

  findings.sort((left, right) => byCodeUnit(left.location.pointer, right.location.pointer) || byCodeUnit(left.ruleId, right.ruleId))

  return {
    schemaVersion: REPORT_SCHEMA_VERSION,
    tool: TOOL_ID,
    status: findings.length > 0 ? 'fail' : 'pass',
    summary: {
      checked: readings.length,
      normalized: normalized.length,
      errors: findings.length,
      warnings: 0,
    },
    readings: normalized,
    findings,
  }
}

export function formatReport(report) {
  const lines = report.findings.map((item) =>
    `${item.severity.toUpperCase().padEnd(7)} ${item.location.pointer} ${item.message}`)
  lines.push('')
  lines.push(
    `${report.summary.checked} reading(s): ${report.summary.normalized} normalized, `
    + `${report.summary.errors} rejected, status ${report.status}.`,
  )
  return `${lines.join('\n')}\n`
}
